Stellar Burger

sadova-nadezhda opened the pull request.

sadova-nadezhda committed:
create project Burger

gennady-bars commented on src/App.tsx:
```diff
@@ -1,35 +1,47 @@
+  useEffect(() => {
+    const fetchIngredients = async () => {
+      const response = await fetch(url);
+      const data = await response.json();
```
Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.


```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

gennady-bars commented on src/App.tsx:
```diff
@@ -1,35 +1,47 @@
+      const response = await fetch(url);
+      const data = await response.json();
+      setIngredients(data.data);
+    };
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

gennady-bars commented on src/App.tsx:
```diff
@@ -1,35 +1,47 @@
+import BurgerIngredients from './components/BurgerIngredients';
+import BurgerConstructor from './components/BurgerConstructor';
+
+import './styles/main.scss';
```
`import './styles/main.scss';
`

Вам нужно использовать **модульные** стили из `App.module.scss`, которые там нужно прописать

gennady-bars commented on src/App.tsx:
```diff
@@ -1,35 +1,47 @@
-import { useState } from 'react'
-import reactLogo from './assets/react.svg'
-import viteLogo from '/vite.svg'
-import './App.css'
```
* `App.tsx` нужно переместить в папку `components`  и там поместить в свою папку `App`. Там же будет  `App.module.scss`

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -0,0 +1,80 @@
+
+export default function BurgerConstructor({data}:BurgerConstructorProps) {
+  const total = 10000;
+  const [isModalOpen, setModalOpen] = useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

gennady-bars commented on src/components/BurgerIngredients/BurgerIngredients.tsx:
```diff
@@ -0,0 +1,156 @@
+  proteins: string;
+  fat: string;
+  carbohydrates: string;
+}
```
Тип для ингредиента повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его  в отдельный файл `types.ts` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.tsx:
```diff
@@ -0,0 +1,80 @@
+  image: string;
+  name: string;
+  price: number;
+}
```
Тип для ингредиента повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его  в отдельный файл `types.ts` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/components/BurgerIngredients/BurgerIngredients.tsx:
```diff
@@ -0,0 +1,156 @@
+            <div className={s.cards}>
+              {data.filter((element) => element.type === 'bun').map((element) => (
+                <IngredientsCard
+                  key={element._id}
```
`                  key={element._id}
`


Отлично, что используете атрибут `key`, и отлично, что для него используете уникальный `_id`, а не индекс массива.

[Почему индекс не нужно использовать, можно прочитать тут]( https://habr.com/ru/companies/hh/articles/352150/)

gennady-bars commented on src/components/BurgerIngredients/BurgerIngredients.tsx:
```diff
@@ -0,0 +1,156 @@
+              {data.filter((element) => element.type === 'sauce').map((element) => (
+                <IngredientsCard
+                  key={element._id}
+                  img={element.image}
+                  name={element.name}
+                  price={element.price}
```
**МОЖНО ЛУЧШЕ**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента

gennady-bars commented on src/components/BurgerIngredients/parts/IngredientsCard/IngredientsCard.tsx:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <div className={s.card} onClick={onClick} >
+      <Counter count={0} size="default" extraClass="m-1" />
+      <div className={classNames(s.card__img)}><img src={img} alt="" /></div>
```
`alt="" ` 

Каждой картинке (тегу `img`) нужен информативный атрибут `alt` (например: имя картинки вставлять), так как при ошибке загрузки картинки на сайте подсветится **название** этой картинки (которое в `alt`). Еще это нужно для **плохо видящих** и `незрячих` людей - для них есть специальные средства, которые распознают текст в `alt` и проговаривают его. Подробнее тут https://habr.com/ru/post/341810/

gennady-bars commented on src/components/IngredientDetails/IngredientDetails.tsx:
```diff
@@ -0,0 +1,44 @@
+export default function IngredientDetails({data}:IngredientDetails) {
+  return (
+    <div className={s.card}>
+      <div className={classNames(s.card__img, "mb-4")}><img src={data.image} alt="" /></div>
```
`alt="" ` 

Каждой картинке (тегу `img`) нужен информативный атрибут `alt` (например: имя картинки вставлять), так как при ошибке загрузки картинки на сайте подсветится **название** этой картинки (которое в `alt`). Еще это нужно для **плохо видящих** и `незрячих` людей - для них есть специальные средства, которые распознают текст в `alt` и проговаривают его. Подробнее тут https://habr.com/ru/post/341810/

gennady-bars commented on src/components/OrderDetails/OrderDetails.tsx:
```diff
@@ -0,0 +1,19 @@
+      <div className={classNames(s.order__number, 'mb-8 text_type_digits-large')}>034536</div>
+      <span className='mb-15 text_type_main-medium'>идентификатор заказа</span>
+      <div className={classNames(s.order__icon, 'mb-15')}>
+        <img src={done} alt="" />
```
`alt="" ` 

Каждой картинке (тегу `img`) нужен информативный атрибут `alt` (например: имя картинки вставлять), так как при ошибке загрузки картинки на сайте подсветится **название** этой картинки (которое в `alt`). Еще это нужно для **плохо видящих** и `незрячих` людей - для них есть специальные средства, которые распознают текст в `alt` и проговаривают его. Подробнее тут https://habr.com/ru/post/341810/

gennady-bars commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,50 @@
+const modalRoot = document.getElementById('modals');
+
+export default function Modal({ title, onClose, children } : ModalProps) {
+  const handleEscClose = (e: KeyboardEvent) => {
```
Обработчик `Escape` нужно поместить внутрь `useEffect` , чтобы не менялась ссылка на функцию при перерисовке компонента

gennady-bars commented on src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,50 @@
+            <CloseIcon type="primary" />
+          </button>
+          {title && (
+            <header className={classNames(s.modal__header, 'mb-5')}>
```
`header` обычно используется для шапки сайта, а тут по семантике уже есть `h2` для заголовка. Нужно просто `div` использовать вместо `header`

gennady-bars commented on src/App.tsx:
```diff
@@ -1,35 +1,47 @@
+    };
+
+    fetchIngredients();
+  }, []);
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Хорошая верстка
* Отлично, что используете атрибут `key`, и отлично, что для него используете уникальный `_id`, а не индекс массива.
* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз
* Хорошая типизация


но есть некоторые недочеты:




*  Попапы **не должны закрываться** при клике **внутри** них, так как внутри могут быть формы, текст, картинки, на которые пользователь будет нажимать, чтобы скопировать и т.д.
* Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.
* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.
* Вам нужно использовать **модульные** стили из `App.module.scss`, которые там нужно прописать
* `App.tsx` нужно переместить в папку `components`  и там поместить в свою папку `App`. Там же будет  `App.module.scss`
* Тип для ингредиента повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его  в отдельный файл `types.ts` в папку `utils` и импортировать туда, где нужно
* Каждой картинке (тегу `img`) нужен информативный атрибут `alt` 
* Обработчик `Escape` нужно поместить внутрь `useEffect` , чтобы не менялась ссылка на функцию при перерисовке компонента


**МОЖНО ЛУЧШЕ**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента

sadova-nadezhda committed:
fixed

App перенесла в components (стилей нет отдельных для него)
Для картинок прописала alt

sadova-nadezhda committed:
Update App.tsx

добавила проверку ответа от сервера

sadova-nadezhda committed:
Update Modal.tsx

Обработчик Escape поместила внутрь useEffect

sadova-nadezhda committed:
update App component

добавила App.module.scss

sadova-nadezhda committed:
Update Modal.tsx

подправила закрытие модалки

sadova-nadezhda committed:
create useModal

создала кастомный хук для управления модальным окном(сложно с ts))

sadova-nadezhda committed:
update container

sadova-nadezhda committed:
Update App.tsx

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
